### PR TITLE
[BugFix] Do not use ColumnViewer::size for regex_match

### LINKED ISSUE #5983: [sqlsmith]mismatched row count: 1 vs 104
### Steps to reproduce the behavior (Required)
1. CREATE TABLE '...'
```
CREATE TABLE `tpcds_100g_promotion` (
  `p_promo_sk` int(11) NOT NULL COMMENT "",
  `p_promo_id` char(16) NOT NULL COMMENT "",
  `p_start_date_sk` int(11) NULL COMMENT "",
  `p_end_date_sk` int(11) NULL COMMENT "",
  `p_item_sk` int(11) NULL COMMENT "",
  `p_cost` decimal64(15, 2) NULL COMMENT "",
  `p_response_target` int(11) NULL COMMENT "",
  `p_promo_name` char(50) NULL COMMENT "",
  `p_channel_dmail` char(1) NULL COMMENT "",
  `p_channel_email` char(1) NULL COMMENT "",
  `p_channel_catalog` char(1) NULL COMMENT "",
  `p_channel_tv` char(1) NULL COMMENT "",
  `p_channel_radio` char(1) NULL COMMENT "",
  `p_channel_press` char(1) NULL COMMENT "",
  `p_channel_event` char(1) NULL COMMENT "",
  `p_channel_demo` char(1) NULL COMMENT "",
  `p_channel_details` varchar(100) NULL COMMENT "",
  `p_purpose` char(15) NULL COMMENT "",
  `p_discount_active` char(1) NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`p_promo_sk`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`p_promo_sk`) BUCKETS 5
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "DEFAULT"
);
CREATE TABLE `colocate_table_2` (
  `k1` datetime NOT NULL COMMENT "",
  `k2` int(11) NOT NULL COMMENT "",
  `v1` double SUM NOT NULL COMMENT ""
) ENGINE=OLAP
AGGREGATE KEY(`k1`, `k2`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`k2`) BUCKETS 8
PROPERTIES (
"replication_num" = "3",
"colocate_with" = "group1",
"in_memory" = "false",
"storage_format" = "DEFAULT"
);
```
2. INSERT INTO '....'
3. SELECT  '....'
```
select  
  regexp(
    cast(from_unixtime(
      cast(unix_timestamp() as INT)) as VARCHAR),
    cast(subq_0.c5 as VARCHAR)) as c0, 
  subq_0.c5 as c1, 
  subq_0.c10 as c2, 
  subq_0.c8 as c3, 
  subq_0.c7 as c4, 
  subq_0.c1 as c5, 
  subq_0.c1 as c6, 
  subq_0.c9 as c7, 
  subq_0.c9 as c8, 
  subq_0.c5 as c9, 
  case when true then case when subq_0.c10 is not NULL then subq_0.c3 else subq_0.c3 end
       else case when subq_0.c10 is not NULL then subq_0.c3 else subq_0.c3 end
       end
     as c10
from 
  (select  
        ref_1.k1 as c0, 
        hll_cardinality(
          cast(ref_0.p_channel_radio as VARCHAR)) as c1, 
        ref_1.k1 as c2, 
        case when ref_0.p_purpose = ref_0.p_channel_details then ref_0.p_cost else ref_0.p_cost end
           as c3, 
        ref_0.p_channel_tv as c4, 
        ref_0.p_promo_id as c5, 
        ref_1.k1 as c6, 
        ref_0.p_purpose as c7, 
        ref_1.v1 as c8, 
        ref_0.p_promo_sk as c9, 
        ref_1.k2 as c10
      from 
        sqlsmith.tpcds_100g_promotion as ref_0
          right join sqlsmith.colocate_table_2 as ref_1
          on (ref_0.p_cost = ref_1.v1 )
      where ref_0.p_cost is not NULL
      limit 106) as subq_0
where case when subq_0.c10 <= seconds_add(
        cast(subq_0.c6 as DATETIME),
        cast(subq_0.c1 as INT)) then subq_0.c5 else subq_0.c5 end
     <> subq_0.c7
limit 141;
```
Unstable results.
106 rows in set (0.05 sec)
104 rows in set (0.04 sec)
ERROR 1064 (HY000): mismatched row count: 1 vs 104
ERROR 1064 (HY000): mismatched row count: 1 vs 106
### Real behavior (Required)
```
ERROR 1064 (HY000): mismatched row count: 1 vs 104
```
```
2022-05-11 19:11:41,590 WARN (starrocks-mysql-nio-pool-11|363) [Coordinator.getNext():847] get next fail, need cancel. status errorCode CANCELLED InternalError, query id: 228e0ef1-d11b-11ec-97b2-00163e0025c6
2022-05-11 19:11:41,591 WARN (starrocks-mysql-nio-pool-11|363) [Coordinator.getNext():868] query failed: mismatched row count: 1 vs 106
```
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
 - BRANCH-2.2-RELEASE 05682ae

### PR DESCRIPTION
**Pull request by ZiheLiu**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #5983.

## Problem Summary：

`regex_match_{partial,full}` uses `ColumnViewer.size` as the number of rows. However, when the column is a constant column, `ColumnViewer.size` returns the size of inner data column size (which is 1) not the logic size.

```C++
ColumnViewer<TYPE_VARCHAR> value_viewer(columns[0]);
for (int row = 0; row < value_viewer.size(); ++row) {
    // ...
}
```

Therefore, don't use `ColumnViewer.size()` for regex_match.

### CODE REVIEW

**Review by murphyatwork — DISMISSED** [review]
LGTM.
If any test cases for it would be better.

**Comment by stdpain** [comment]
It may cause executing multi times for some unary function such as cast.
we shouldn't call value_viewer.size() in no-unary function

File: be/src/column/column_helper.h [hunk 1]
```diff
@@ -265,6 +265,13 @@ class ColumnHelper {
 
     static bool is_all_const(const Columns& columns);
 
+    // Returns
+    //  1. whether all the columns are constant.
+    //  2. the number of the packed rows. If all the columns are constant, it will be 1,
+    //     which could reduce unnecessary calculations.
+    //     Don't forget to resize the result constant columns if necessary.
+    static pair<bool, size_t> num_packed_rows(const Columns& columns);
```
Comment (stdpain) [inline be/src/column/column_helper.h:273]: prefer std::pair

File: be/src/column/column_helper.cpp [hunk 2]
```diff
@@ -240,6 +240,19 @@ bool ColumnHelper::is_all_const(const Columns& columns) {
     return std::all_of(std::begin(columns), std::end(columns), [](const ColumnPtr& col) { return col->is_constant(); });
 }
 
+pair<bool, size_t> ColumnHelper::num_packed_rows(const Columns& columns) {
```
Comment (kangkaisen) [inline be/src/column/column_helper.cpp]: We needn't this method.   we could always use columns[0] size to get the chunk row number.

Comment (stdpain) [reply be/src/column/column_helper.cpp]: I think this method is very useful and could solve many problems.

**Review by stdpain — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Pull request merged**